src/test/script_tests.cpp

```diff
@@ -1355,6 +1355,192 @@ BOOST_AUTO_TEST_CASE(script_GetScriptAsm)
+
+    stack.clear();
+    CScript scriptBadSigType = CScript() << data << sigbadtype;
+    BOOST_CHECK(EvalScript(stack, scriptBadSigType, 0, sigChecker, &serror, nullptr));
+    BOOST_CHECK(!EvalScript(stack, condScript, 0, sigChecker, &serror, nullptr));
+
+    stack.clear();
+    sigtype[2] ^= 1; // Screw up the signature
+    proveScript = CScript() << data << sigtype;
+    BOOST_CHECK(EvalScript(stack, proveScript, 0, sigChecker, &serror, nullptr));
+    BOOST_CHECK(!EvalScript(stack, condScript, 0, sigChecker, &serror, nullptr));
+    sigtype[2] ^= 1; // back to correct sigtype
+
+    QuickAddress u2;
+
+    // Now try a more realistic script
+    // Here I am validating that "dataSigner" signed a piece of data that is equal a
+    // particular value, and that that transaction is signed by the txo address in the
+    // normal p2pkh fashion.  In a real transaction, you would likely use string instructions
+    // to break the data into pieces, such as <ticker>, <date>, and <price> if importing information
+    // about a security.  However, these string instructions are not yet enabled.
+    condScript = CScript() << ToByteVector(dataSigner.addr) << OP_DATASIGVERIFY << data << OP_EQUALVERIFY << OP_DUP
+                           << OP_HASH160 << ToByteVector(u2.addr) << OP_EQUALVERIFY << OP_CHECKSIG;
+    BOOST_CHECK(condScript.GetSigOpCount(true) == 2);
+    unsigned int sighashType = SIGHASH_ALL | SIGHASH_FORKID;
+    std::vector<unsigned char> txoSig;
+    // Since I don't have a tx, I'm going to use a fake tx hash, which is just the hash
+    // of the public key.
+    // uint256 hash = SignatureHash(prevtx.vout[prevout].scriptPubKey, tx, 0, sighashType, prevtx.vout[prevout].nValue,
+    // 0);
+    uint256 hash = u2.pubkey.GetHash();
+    if (!u2.secret.Sign(hash, txoSig))
+    {
+        assert(0);
+    }
+    txoSig.push_back((unsigned char)sighashType);
+    SigPubkeyHashChecker pkhChecker;
+
+    proveScript = CScript() << txoSig << ToByteVector(u2.secret.GetPubKey()) << data << sigtype;
+
+    stack.clear();
+    BOOST_CHECK(EvalScript(stack, proveScript, 0, pkhChecker, &serror, nullptr));
+    BOOST_CHECK(EvalScript(stack, condScript, 0, pkhChecker, &serror, nullptr));
+
+    stack.clear();
+    sigtype[3] ^= 1; // Screw up the data signature
+    proveScript = CScript() << data << sigtype << ToByteVector(u2.pubkey);
+    BOOST_CHECK(EvalScript(stack, proveScript, 0, pkhChecker, &serror, nullptr));
+    BOOST_CHECK(!EvalScript(stack, condScript, 0, pkhChecker, &serror, nullptr));
+    sigtype[3] ^= 1; // back to correct data sig
+    // provide the wrong utxo pubkey
+    proveScript = CScript() << data << sigtype << ToByteVector(dataSigner.pubkey);
+    stack.clear();
+    BOOST_CHECK(EvalScript(stack, proveScript, 0, pkhChecker, &serror, nullptr));
+    BOOST_CHECK(!EvalScript(stack, condScript, 0, pkhChecker, &serror, nullptr));
+
+    enableDataSigVerify = priorDataSigVerifyValue;
+}
+
```

Comment:
I can't tell from reading through the comments in the tests whether you covered all the  cases for when the script should fail as taken from you spec as follows: 

"If there are less then 3 items on the stack, the script fails.  If the pubkeyhash field is not 20 bytes, the script fails.  If the *'type and signature'* field is not 66 bytes, the script fails.  If the first byte of the *'type and signature'* field is not DATASIG_COMPACT_ECDSA (1), the script fails."

It would be good to see different sections in the tests...such as tests for failures/attacks , and tests for real scenarios that should pass.  I think having a really solid set of tests will make this feature much more acceptable to anyone who has doubts about its safety:

Reply:
I went through and added comments and additional tests to check each error condition.